### PR TITLE
[SCRIPTS] XLNet squad finetuning scripts

### PR DESCRIPTION
**Pull request by zburning**
## Description ##
finetuning squad

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1012/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/1/index.html

**Comment by mli** [comment]
Job PR-1012/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/2/index.html

File: scripts/language_model/model/qa.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            end_logits = self.end_logits(
+                hidden_states_expanded, start_states=start_states)  # shape (bsz, slen, start_n_top)
+            end_log_probs = mx.nd.softmax(end_logits, axis=1)  # shape (bsz, slen, start_n_top)
+            end_top_log_probs, end_top_index = mx.ndarray.topk(
+                end_log_probs, k=self.end_top_n, axis=1,
+                ret_typ='both')  # shape (bsz, end_n_top, start_n_top)
+            end_top_log_probs = end_top_log_probs.reshape((-1, self.start_top_n * self.end_top_n))
+            end_top_index = end_top_index.reshape((-1, self.start_top_n * self.end_top_n))
+
+            #einsum("blh,bl->bh", output, start_log_probs) need further check
```
Comment (leezu) [inline scripts/language_model/model/qa.py]: > need further check

?

File: scripts/language_model/model/qa.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        self.layernorm = nn.LayerNorm()
+
+    def __call__(self, hidden_states, start_states=None, start_positions=None, p_masks=None):
+        # pylint: disable=arguments-differ
+        return super(PoolerEndLogits, self).__call__(hidden_states, start_states, start_positions,
+                                                     p_masks)
+
+    def hybrid_forward(self, F, hidden_states, start_states, start_positions, p_mask):
+        # pylint: disable=arguments-differ
+        assert start_states is not None or start_positions is not None
```
Comment (leezu) [inline scripts/language_model/model/qa.py]: Should this rather be XOR (one and only one should be not None)?

Note that due to limitation of current HybridBlock, you can't change the argument that is `None` in subsequent calls after hybridization. The computational graph will be fixed.
Does this currently work for you? My understanding is that you use `start_positions` during training and `start_states` during testing? Did you run the script with `args.only_predict is True`?

File: scripts/language_model/model/qa.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            first sentence or the second one.
+        valid_length : NDArray or None, shape (batch_size,)
+            Valid length of the sequence. This is used to mask the padded tokens.
+
+        Returns
+        -------
+        outputs : NDArray
+            Shape (batch_size, seq_length, 2)
+        """
+        # XXX Temporary hack for hybridization as hybridblock does not support None inputs
```
Comment (leezu) [inline scripts/language_model/model/qa.py]: Not needed anymore, see #967

**Comment by mli** [comment]
Job PR-1012/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/3/index.html

**Comment by mli** [comment]
Job PR-1012/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/4/index.html

**Comment by mli** [comment]
Job PR-1012/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/5/index.html

**Comment by mli** [comment]
Job PR-1012/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/6/index.html

**Comment by mli** [comment]
Job PR-1012/29 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/29/index.html

**Comment by mli** [comment]
Job PR-1012/30 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/30/index.html

File: scripts/language_model/run_squad.py [hunk 4]
```diff
@@ -82,7 +82,7 @@
     '--uncased',
     action='store_true',
     help=
-    'if set, inputs are converted to lower case. Up to 01/04/2020, all released models are cased'
+    'if set, inputs are converted to lower case. Up to 04/01/2020, all released models are cased'
```
Comment (leezu) [inline scripts/language_model/run_squad.py]: Some countries default to MM/DD, others DD/MM. Let's use January 4th 2020 to be clear

**Comment by mli** [comment]
Job PR-1012/31 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1012/31/index.html

**Comment by zburning** [comment]
Including this in [#1124](https://github.com/dmlc/gluon-nlp/pull/1124)

**Pull request closed without merging**